### DISCUSSION TITLE
How to capture rust stdout output from python

### DISCUSSION THREAD

**Discussion by RobertCraigie** [PyO3/pyo3#1960] [Questions]
I've tried using pytest's [capsys](https://docs.pytest.org/en/6.2.x/capture.html) fixture and python's [contextlib.redirect_stdout](https://docs.python.org/3/library/contextlib.html#contextlib.redirect_stdout) but neither seemed to capture any output.

Is capturing stdout output from python supported?

Thank you for this fantastic library!

**Comment by davidhewitt** [comment]
See https://github.com/PyO3/pyo3/discussions/1918#discussioncomment-1473356

**Comment by tcztzy** [comment]
I am in a more sophisticated situation, I use `autocxx` to wrapper C++ codes and the stdout is come from C++ code. But I can give some suggestions for you.

You can impl the `std::io::Write` trait for a rust struct

```rust
use std::io::Write;
struct PySysStdout;

impl Write for PySysStdout {
    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
        let s = std::str::from_utf8(buf).unwrap();
        Python::with_gil(|py| {
            let locals = PyDict::new(py);
            locals.set_item("s", s).unwrap();
            py.run("print(s, end='')", None, Some(&locals)).unwrap();
        });
        Ok(buf.len())
    }
    fn flush(&mut self) -> std::io::Result<()> {
        Python::with_gil(|py| {
            py.run("import sys;sys.stdout.flush()", None, None).unwrap();
        });
        Ok(())
    }
}
```

And then, refer [this StackOverflow answer](https://stackoverflow.com/a/72187752/5434822).

Hope it works for you!
